src/GitHub.Api/Authentication/LoginManager.cs

```diff
@@ -83,108 +73,59 @@ class LoginManager : ILoginManager
-                EnsureNonNullAuthorization(auth);
-            }
-            catch (TwoFactorAuthorizationException e)
-            {
-                LoginResultCodes result;
-                if (e is TwoFactorRequiredException)
+                token = await TryLogin(host, username, password);
+                if (string.IsNullOrEmpty(token))
                 {
-                    result = LoginResultCodes.CodeRequired;
-                    logger.Trace("2FA TwoFactorAuthorizationException: {0} {1}", LoginResultCodes.CodeRequired, e.Message);
+                    throw new InvalidOperationException("Returned token is null or empty");
                 }
-                else
-                {
-                    result = LoginResultCodes.CodeFailed;
-                    logger.Error(e, "2FA TwoFactorAuthorizationException: {0} {1}", LoginResultCodes.CodeRequired, e.Message);
-                }
-
-                return new LoginResultData(result, e.Message, host, newAuth);
-            }
-            catch(LoginAttemptsExceededException e)
-            {
-                logger.Warning(e, "Login LoginAttemptsExceededException: {0}", e.Message);
-
-                await keychain.Clear(host, false);
-                return new LoginResultData(LoginResultCodes.LockedOut, Localization.LockedOut, host);
             }
-            catch (ApiValidationException e)
+            catch (TwoFactorRequiredException e)
             {
-                logger.Warning(e, "Login ApiValidationException: {0}", e.Message);
+                LoginResultCodes result;
+                result = LoginResultCodes.CodeRequired;
+                logger.Trace("2FA TwoFactorAuthorizationException: {0} {1}", LoginResultCodes.CodeRequired, e.Message);
 
-                var message = e.ApiError.FirstErrorMessageSafe();
-                await keychain.Clear(host, false);
-                return new LoginResultData(LoginResultCodes.Failed, message, host);
+                return new LoginResultData(result, e.Message, host, password);
             }
             catch (Exception e)
             {
                 logger.Warning(e, "Login Exception");
 
-                // Some enterprise instances don't support OAUTH, so fall back to using the
-                // supplied password - on instances that don't support OAUTH the user should
-                // be using a personal access token as the password.
-                if (EnterpriseWorkaround(host, e))
-                {
-                    auth = new ApplicationAuthorization(password);
-                }
-                else
-                {
-                    await keychain.Clear(host, false);
-                    return new LoginResultData(LoginResultCodes.Failed, Localization.LoginFailed, host);
-                }
+                await keychain.Clear(host, false);
+                return new LoginResultData(LoginResultCodes.Failed, Localization.LoginFailed, host);
```

Comment:
This is missing the other error conditions like locked out and the workaround for enterprise, isn't it?

Reply:
Tracked in #617